cli/src/main/java/org/jboss/sbomer/cli/feature/sbom/processor/DefaultProcessor.java

```diff
@@ -266,9 +275,102 @@ public Bom process(Bom bom) {
+            return;
+        }
+        String buildID = parseBuildIdFromURL(externalReferences.get(0).getUrl());
+        if (buildID == null) {
+            log.warn("Could not parse PNC build ID from url {}", externalReferences.get(0).getUrl());
+            return;
+        }
+        Build build = pncService.getBuild(buildID);
+        if (build == null) {
+            log.warn("Could not obtain PNC build for build id {}", buildID);
+            return;
+        }
+        if (build.getBuildConfigRevision().getBuildType() == BuildType.NPM) {
+            // Not processing pure NPM builds, cyclonedx-nodejs plugin handles them correctly
+            return;
+        }
+        Collection<Artifact> npmDependencies = pncService.getNPMDependencies(buildID);
+        if (npmDependencies.isEmpty()) {
+            // No NPM dependencies to add
+            return;
+        }
+
+        addMissingNpmDependencies(bom, npmDependencies, build.getBuildConfigRevision().getBuildType());
+    }
+
+    private void addMissingNpmDependencies(Bom bom, Collection<Artifact> npmDependencies, BuildType buildType) {
+        Set<String> listedPurls = bom.getComponents().stream()
+            .map(DefaultProcessor::getPackageURL)
+            .map(PackageURL::getCoordinates)
+            .collect(Collectors.toSet());
+
+        Set<String> dependencyRefs = new HashSet<>();
+        for (Artifact artifact : npmDependencies) {
+            String coordinates;
+            try {
+                coordinates = new PackageURL(artifact.getPurl()).getCoordinates();
+            } catch (MalformedPackageURLException e) {
+                log.warn("Unable to parse artifact purl '{}'.", artifact.getPurl());
+                continue;
+            }
+            if (listedPurls.contains(coordinates)) {
+                continue;
+            }
+            Component newComponent = createComponent(artifact, Component.Scope.REQUIRED, Component.Type.LIBRARY, buildType);
+            setArtifactMetadata(newComponent, artifact, pncService.getApiUrl());
+            setPncBuildMetadata(newComponent, artifact.getBuild(), pncService.getApiUrl());
+            bom.addComponent(newComponent);
+            dependencyRefs.add(newComponent.getBomRef());
+        }
+        addMainComponentDependencies(bom, dependencyRefs);
+    }
+
+    private void addMainComponentDependencies(Bom bom, Set<String> dependencies) {
+        String mainComponentRef = bom.getMetadata().getComponent().getBomRef();
+
+        Dependency mainComponentDependencies = bom.getDependencies().stream()
+            .filter(d -> mainComponentRef.equals(d.getRef()))
+            .findFirst()
+            .orElseGet(() -> addNewDependency(bom, mainComponentRef));
+
+        dependencies.stream().map(SbomUtils::createDependency).forEach(mainComponentDependencies::addDependency);
```

Comment:
From what I understand, this is creating a new dependency and adding it to the mainComponentDependencies (so it ends up in the dependsOn[]), but should it also add the new dependency to the bom? So that each dependency is also listed singularly (with an empty dependsOn[]) ?
I guess I want to say, should we also have a `bom.addDependency(newDep)` for each newly created dependency?

Reply:
I think you missed the line I highlighted here: https://github.com/project-ncl/sbomer/pull/993/files#r1866566103

Reply:
I guess what I meant is that I cannot see where each new dependency is added to the bom. 
For every NPM artifact:
* I see that a new Component is added to the BOM
* I see that a new Dependency is created and that dependency is added to the main component dependency list
* I do not see that the new Dependency is also added to the BOM. Something like:
 
```
        dependencies.stream().map(SbomUtils::createDependency).forEach(d -> {
            mainComponentDependencies.addDependency(d);
            bom.addDependency(d);
        });
```
The original dependency section of the BOM for the main component is:
```
    {
      "ref": "pkg:maven/org.keycloak/keycloak-parent@24.0.6.redhat-00001?type=pom",
      "dependsOn": [
        "pkg:maven/org.keycloak/keycloak-common@24.0.6.redhat-00001?type=jar"
      ]
    }
```

With this PR original code, this section changed in the processed BOM:
```
    {
      "ref": "pkg:maven/org.keycloak/keycloak-parent@24.0.6.redhat-00001?type=pom",
      "dependsOn": [
        "pkg:maven/org.keycloak/keycloak-common@24.0.6.redhat-00001?type=jar",
        "pkg:npm/%40redhat/kogito-tooling-keyboard-shortcuts@0.9.0-2",
        "pkg:npm/once@1.4.0"
      ]
    }
```
and nothing else. I would have expected also 2 more dependencies. With my suggestion ^, these are the changes:
```
    {
      "ref": "pkg:maven/org.keycloak/keycloak-parent@24.0.6.redhat-00001?type=pom",
      "dependsOn": [
        "pkg:maven/org.keycloak/keycloak-common@24.0.6.redhat-00001?type=jar",
        "pkg:npm/%40redhat/kogito-tooling-keyboard-shortcuts@0.9.0-2",
        "pkg:npm/once@1.4.0"
      ]
    },
    {
      "ref": "pkg:npm/%40redhat/kogito-tooling-keyboard-shortcuts@0.9.0-2",
      "dependsOn": []
    },
    {
      "ref": "pkg:npm/once@1.4.0",
      "dependsOn": []
    }
```
which I think are more correct for the hierarchy consistency.